Sources/SwiftParserDiagnostics/ParseDiagnosticsGenerator.swift

```diff
@@ -171,26 +171,38 @@ public class ParseDiagnosticsGenerator: SyntaxAnyVisitor {
     return handledNodes.contains(node.id)
   }
 
-  /// Utility function to emit a diagnostic that removes a misplaced token and instead inserts an equivalent token at the corrected location.
+  /// Utility function to emit a diagnostic that removes a misplaced node and instead inserts an equivalent node at the corrected location.
   ///
-  /// If `incorrectContainer` contains some tokens that satisfy `unexpectedTokenCondition`, emit a diagnostic with message `message` that marks this token as misplaced.
-  /// If `correctTokens` contains missing tokens, also emit a Fix-It with message `fixIt` that marks the unexpected token as missing and instead inserts `correctTokens`.
-  public func exchangeTokens(
+  /// If `incorrectContainer` contains some nodes that satisfy `unexpectedNodeCondition`, emit a diagnostic with
+  /// message `message` that marks the present tokens in this node as misplaced.
+  ///
+  /// If `correctNodes` contains missing tokens, also emit a Fix-It with message `fixIt` that marks the present tokens
+  /// in unexpected nodes as missing while inserting the missing tokens in `correctNodes`.
+  public func exchangeNodes<S: SyntaxProtocol>(
     unexpected: UnexpectedNodesSyntax?,
-    unexpectedTokenCondition: (TokenSyntax) -> Bool,
-    correctTokens: [TokenSyntax?],
+    unexpectedNodeCondition: (S) -> Bool,
+    correctNodes: [S?],
     message: (_ misplacedTokens: [TokenSyntax]) -> some DiagnosticMessage,
     moveFixIt: (_ misplacedTokens: [TokenSyntax]) -> FixItMessage,
     removeRedundantFixIt: (_ misplacedTokens: [TokenSyntax]) -> FixItMessage? = { _ in nil }
   ) {
     guard let incorrectContainer = unexpected else { return }
-    let misplacedTokens = incorrectContainer.presentTokens(satisfying: unexpectedTokenCondition)
-    if misplacedTokens.isEmpty { return }
+    let misplacedNodes = incorrectContainer.compactMap { $0.as(S.self) }.filter(unexpectedNodeCondition)
+    let misplacedTokens = misplacedNodes.flatMap {
+      Array($0.tokens(viewMode: .sourceAccurate))
+    }
+    guard !misplacedTokens.isEmpty else { return }
 
-    let correctTokens = correctTokens.compactMap({ $0 })
+    let correctTokens = correctNodes.flatMap { node -> [TokenSyntax] in
+      if let node {
+        Array(node.tokens(viewMode: .all))
+      } else {
+        []
+      }
+    }
```

Comment:
I think this could also be `[Syntax]`:
- `correctTokens.only` would just need to become `correctTokens.only.as(TokenSyntax.self)`
- `correctTokens.filter(\.isMissing)` could become `correctTokens.filter(\.isMissingAllTokens)`